cpu/atmega_common/periph/uart.c

```diff
@@ -178,8 +172,9 @@ void uart_write(uart_t uart, const uint8_t *data, size_t len)
 #endif
         /* start of TX won't finish until no data in UDRn and transmit shift
            register is empty */
-        _tx_pending |= (1 << uart);
-
+        unsigned long state = irq_disable();
+        atmega_state |= ATMEGA_STATE_FLAG_UART_TX(uart);
+        irq_restore(state);
```

Comment:
This is the point I expect to increase the ROM. The issue is that the AVR cannot set a flag in SRAM atomically. The original version would have looked like:

```asm
lds r24, _tx_pending
or r24, r25
sts _tx_pending, r24
```

Or in my own words:

1. Load `_tx_pending` into a register
2. Set a bit in the register (according to the value in `uart`)
3. Write the register back into `_tx_pending`

Lets say in `_tx_pending` the bit for UART1 is set, and for UART0 is not. Now command 1. is executed and this state is loaded into the register. If the IRQ kicks in and clears the bit in `_tx_pending` this update is lost, as command 3 will restore this bit. Therefore this whole "load, set bit, store" sequence needs to be done atomically. The AVR cannot do this in a single instruction, so we have to resort to disabling interrupts. (Or using C11 atomic, which would internally do the same but with more overhead.)